heft-plugins/heft-lint-plugin/src/SarifFormatter.ts

```diff
@@ -0,0 +1,264 @@
+    tags?: string[];
+    problem?: {
+      severity?: 'recommendation' | 'warning' | 'error';
+      securitySeverity?: number;
+    };
+  };
+}
+
+export interface ISarifFile {
+  location: {
+    uri: string;
+  };
+}
+
+export interface ISuppressedAnalysis {
+  kind: string;
+  justification: string;
+}
+
+export interface ISarifLocation {
+  physicalLocation: ISarifPhysicalLocation;
+}
+
+export interface ISarifPhysicalLocation {
+  artifactLocation: {
+    uri: string;
+    index: number;
+  };
+  region?: IRegion;
+}
+
+interface IMessage extends TEslint.Linter.LintMessage {
+  suppressions?: ISuppressedAnalysis[];
+}
+
+const INTERNAL_ERROR_ID: 'ESL0999' = 'ESL0999';
+const SARIF_VERSION: '2.1.0' = '2.1.0';
+const SARIF_INFORMATION_URI: 'http://json.schemastore.org/sarif-2.1.0-rtm.5' =
+  'http://json.schemastore.org/sarif-2.1.0-rtm.5';
+/**
+ * Converts ESLint results into a SARIF (Static Analysis Results Interchange Format) log.
+ *
+ * This function takes in a list of ESLint lint results, processes them to extract
+ * relevant information such as errors, warnings, and suppressed messages, and
+ * outputs a SARIF log which conforms to the SARIF v2.1.0 specification.
+ *
+ * @param results - An array of lint results from ESLint that contains linting information,
+ *                  such as file paths, messages, and suppression details.
+ * @param options - An object containing options for formatting:
+ *                  - `ignoreSuppressed`: Boolean flag to decide whether to ignore suppressed messages.
+ *                  - `eslintVersion`: Optional string to include the version of ESLint in the SARIF log.
+ * @returns The SARIF log containing information about the linting results in SARIF format.
+ */
+export function formatEslintResultsAsSARIF(
+  results: TEslint.ESLint.LintResult[],
+  options: ISerifFormatterOptions
+): ISarifLog {
+  const { ignoreSuppressed, eslintVersion } = options;
+  const toolConfigurationNotifications: ISarifRepresentation[] = [];
+  const sarifFiles: Map<string, ISarifFile> = new Map();
+  const sarifResults: ISarifRepresentation[] = [];
+
+  const sarifRun: ISarifRun = {
+    tool: {
+      driver: {
+        name: 'ESLint',
+        informationUri: 'https://eslint.org',
+        version: eslintVersion,
+        rules: []
+      }
+    }
+  };
+
+  const sarifLog: ISarifLog = {
+    version: SARIF_VERSION,
+    $schema: SARIF_INFORMATION_URI,
+    runs: [sarifRun]
+  };
+
+  let executionSuccessful: boolean = true;
+
+  for (const result of results) {
+    const { filePath } = result;
+    const artifactIndex: number = sarifFiles.size;
+    const fileUrl: string = url.pathToFileURL(filePath).toString();
+    let sarifFile: ISarifFile | undefined = sarifFiles.get(filePath);
+
+    if (sarifFile === undefined) {
+      sarifFile = {
+        location: {
+          uri: fileUrl
+        }
+      };
+      sarifFiles.set(filePath, sarifFile);
+    }
```

Comment:
```suggestion
    let sarifFile: ISarifFile | undefined = sarifFiles.get(filePath);

    if (sarifFile === undefined) {
      const fileUrl: string = url.pathToFileURL(filePath).toString();
      sarifFile = {
        location: {
          uri: fileUrl
        }
      };
      sarifFiles.set(filePath, sarifFile);
    }
```

As written , the `artifactIndex` value won't be correct if a file has multiple issues. You'll need to track the artifact index separately. 

Keep a map of file paths to artifact indicies in addition to a normal array of `ISarifFile` objects.

Reply:
`artifactIndex` is now properly tracked.